_posts/2016-09-10-akka-stream-kafka.md

```diff
@@ -0,0 +1,65 @@
+
+```scala
+val consumerSettings = ConsumerSettings(system, 
+  new ByteArrayDeserializer, new StringDeserializer)
+  .withBootstrapServers("localhost:9092")
+  .withGroupId("group1")
+  .withProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+
+val result = Consumer
+  .committableSource(consumerSettings, Subscriptions.topics("topic1"))
+  .mapAsync(3)(processMsg)
+  .groupedWithin(30, 10.seconds) // commit every 10 seconds or 30 msgs
+  .map(group => group.foldLeft(CommittableOffsetBatch.empty) { 
+    (batch, elem) => batch.updated(elem) })
+  .mapAsync(3)(_.commitScaladsl())
+  .runWith(Sink.ignore)
+
+result.onFailure {
+  case e: Throwable => handleError(e)
+}
+```
+
+Secondly, there are significant performance improvements. We’ve done some benchmarks, and while there is still some overhead from the reactive-kafka wrapper (though keep in mind, that you also get something in return: e.g. all of the benefits coming from backpressure!), the overall numbers look very good and are getting better.
+
+Here’s a couple of common Kafka usage scenarios, comparing the old reactive-kafka version (M4), the current version (0.11), and equivalent functionality implemented using plain Kafka `Producer`s/`Consumer`s (but not taking into account, of course, connecting with any other reactive components).
+
+![Plain consumer]({{ site.url }}/assets/kafka-bench1.png)
+
+The first scenario represents a consumer which reads messages from Kafka and pushes it through a non-blocking processing stage, without commit. The reactive setup (akka-plain-consumer) gets very close (80%) to full speed achieved with a while loop pulling elements from a consumer (plain-consumer).
+
+![Consumer with batched commit]({{ site.url }}/assets/kafka-bench2.png)
+
+Another scenario represents a very common use case - consuming messages and committing in batches in order to achieve at-least once delivery. Performance improvements in 0.11 allowed to gain a massive throughput increase from ~17k to ~450k msgs/s.
+
+![At most once consumer]({{ site.url }}/assets/kafka-bench3.png)
+
+Sometimes one needs to commit each single message before processing, which gives at most once delivery guarantee. 0.11 optimizations fixed the super-slow value of 20 msgs/s and reached over 2800 msgs/s. This is a great step forward, very close to the simple while loop scenario (“at-most-once”).
+
+![Producer]({{ site.url }}/assets/kafka-bench4.png)
+
+Our last benchmark tests a producer continuously sending messages to a Kafka topic. Current version of akka-stream-kafka writes over 85,000 messages per second. This result has been achieved with producer parallelism set to 100, which indicates how many parallel writes can be waiting for confirmation until this stage backpressures. The “plain-producer” test chosen for comparison writes to Kafka in a loop being bound only by its internal buffer size.
+
+Give it a try, we’d love to hear your feedback!
+
+
```

Comment:
@adamw @kciesielski just FYI, I guess it's too late to fix that... The diagrams look not as good, they make one focus that we're slower, and forget about the benefits.
Competition is saying random things and numbers claiming they're "faster than native", which ofc is not true, but from a selling the awesomeness of the lib perspective we're not doing too well here.

I guess it does not matter as much, anyone who tried to use the nativ APIs knows it's a pain and will like the Akka Streams abstraction